tests/dns/dns-rcode/test.yaml

```diff
@@ -0,0 +1,36 @@
+requires:
+  min-version: 8
+
+pcap: input.pcap
+
+checks:
+  - filter:
+      count: 1
+      match:
+        alert.signature_id: 1
+        direction: to_client
+        app_proto: dns
+        dns.rcode: 12
```

Comment:
Upon further investigation on the pcap suggested by Jason for this test (https://github.com/OISF/suricata-verify/blob/master/tests/dns-eve-v2-udp-nxdomain-soa/dns-udp-nxdomain-soa.pcap, right?) I could only find dns.rcode==3 in it, maybe that's what's causing this sv test to fail?

Possible tip: on Wireshark, you can use the `dns.flags.rcode` filter to match on rcode values.

Reply:
I ran the `dns.flags.rcode` and you're right there's only one `rcode` value I see i.e: 
`.... .... .... 0011 = Reply code: No such name (3)`. 
So should I alter all the tests rules to `dns.rcode: 3`?

Reply:
What do you see in the log output? This is the next debug step! ;)

Reply:
I changed the tests to `dns.rcode:3` and I ran `cat output/eve.json | jq 'select(.dns) | .dns' output/eve.json` for which I get this output. The tests are still failing. Should I create a new PR so it's easier for you to code review?
```
{
  "type": "query",
  "id": 33429,
  "rrname": "dne.oisf.net",
  "rrtype": "A",
  "tx_id": 0,
  "opcode": 0
}
{
  "answer": {
    "version": 2,
    "type": "answer",
    "id": 33429,
    "flags": "8183",
    "qr": true,
    "rd": true,
    "ra": true,
    "opcode": 0,
    "rrname": "dne.oisf.net",
    "rrtype": "A",
    "rcode": "NXDOMAIN",
    "authorities": [
      {
        "rrname": "oisf.net",
        "rrtype": "SOA",
        "ttl": 899,
        "soa": {
          "mname": "ns-110.awsdns-13.com",
          "rname": "awsdns-hostmaster.amazon.com",
          "serial": 1,
          "refresh": 7200,
          "retry": 900,
          "expire": 1209600,
          "minimum": 86400
        }
      }
    ]
  }
}
{
  "answer": {
    "version": 2,
    "type": "answer",
    "id": 33429,
    "flags": "8183",
    "qr": true,
    "rd": true,
    "ra": true,
    "opcode": 0,
    "rrname": "dne.oisf.net",
    "rrtype": "A",
    "rcode": "NXDOMAIN",
    "authorities": [
      {
        "rrname": "oisf.net",
        "rrtype": "SOA",
        "ttl": 899,
        "soa": {
          "mname": "ns-110.awsdns-13.com",
          "rname": "awsdns-hostmaster.amazon.com",
          "serial": 1,
          "refresh": 7200,
          "retry": 900,
          "expire": 1209600,
          "minimum": 86400
        }
      }
    ]
  }
}
{
  "version": 2,
  "type": "answer",
  "id": 33429,
  "flags": "8183",
  "qr": true,
  "rd": true,
  "ra": true,
  "opcode": 0,
  "rrname": "dne.oisf.net",
  "rrtype": "A",
  "rcode": "NXDOMAIN",
  "authorities": [
    {
      "rrname": "oisf.net",
      "rrtype": "SOA",
      "ttl": 899,
      "soa": {
        "mname": "ns-110.awsdns-13.com",
        "rname": "awsdns-hostmaster.amazon.com",
        "serial": 1,
        "refresh": 7200,
        "retry": 900,
        "expire": 1209600,
        "minimum": 86400
      }
    }
  ]
}
```

Reply:
For now, this works :)
Thanks for sharing the output! Considering how the SV tests work, do you want to try to guess why the checks are failing, even though you've changed them to `dns.rcode:3`?

Reply:
I've been looking into this for quite a bit now and I've tried other values and I tried changing the rules too but I get the same failures. I don't know if there's something wrong with the schema because this is the error I get: 

```
    Finished release [optimized] target(s) in 0.17s
===> dns-rcode: Sub test #1: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'alert.signature_id': 1, 'direction': 'to_client', 'app_proto': 'dns', 'dns.rcode': 3}}
===> dns-rcode: Sub test #2: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'alert.signature_id': 1, 'direction': 'to_server', 'app_proto': 'dns', 'dns.rcode': 3}}
===> dns-rcode: Sub test #3: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'alert.signature_id': 2, 'direction': 'to_server', 'app_proto': 'dns', 'dns.rcode': 3}}
===> dns-rcode: Sub test #4: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'alert.signature_id': 3, 'direction': 'to_client', 'app_proto': 'dns', 'dns.rcode': 3}}
===> dns-rcode: Sub test #5: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'alert.signature_id': 4, 'direction': 'to_client', 'app_proto': 'dns', 'dns.negate': True, 'dns.rcode': 3}}
/Users/hadiqa/suricata-verify/tests/dns/dns-rcode/output/eve.json - INVALID. Errors:
1./dns/answer Additional properties are not allowed ('authorities' was unexpected)
/Users/hadiqa/suricata-verify/tests/dns/dns-rcode/output/eve.json - INVALID. Errors:
1./dns/answer Additional properties are not allowed ('authorities' was unexpected)
===> dns-rcode: FAILED: Invalid JSON schema
```

Reply:
So, we have two different things going on.
The `Invalid JSON schema` message is expected, as `rcode is not present there yet`, so the json schema validator complains. 
This is a step that will have to be done in one of your Suricata PR iterations - once the `rcode` type is clearly defined. Here's an example of a commit that updates the JSON schema: https://github.com/OISF/suricata/commit/3802a515521151617a

The SV checks failures (Sub test #... FAIL) are related to suricata verify expecting to see `dns.rcode:3` in the eve.json log, but instead, what do we have there? Check https://github.com/OISF/suricata/blob/master/rust/src/dns/log.rs#L513 :P

Reply:
So should I change the log.rs code for `rcode` to be similar to `opcode` as both of these fields are quite similar?

Reply:
imho, yes, would be easier for users than to know how we have defined the codes to be, in string form.
@jasonish what do you say, here?

Reply:
I'm not sure I follow, we didn't change the logging did we?  Chaning the logging would be a breaking change that requires careful consideration.

Reply:
The rcode sv tests are failing even though I'm using `rcode:3` in the rule which is the value for rcode in the pcap file here: `./tests/dns-eve-v2-udp-nxdomain-soa/dns-udp-nxdomain-soa.pcap`. I'm not sure why this is happening. I haven't changed the rust/dns/log.rs file.

Reply:
My guess is you have some local changes to `suricata/etc/schema.json` or you're missing some latest code as `authorities` already exists in the schema.
Clean your repo, rebase your changes on top of latest master and try again please.